server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -249,14 +250,19 @@ public int getNumReducePhases() {
 
         @Override
         public synchronized void close() {
-            assert hasPendingMerges() == false : "cannot close with partial reduce in-flight";
+            if (hasPendingMerges()) {
+                throw new IllegalStateException("Attempted to close with partial reduce in-flight");
+            }
             if (hasFailure()) {
                 assert circuitBreakerBytes == 0;
                 return;
             }
             assert circuitBreakerBytes >= 0;
             circuitBreaker.addWithoutBreaking(-circuitBreakerBytes);
             circuitBreakerBytes = 0;
+            Releasables.close(
+                buffer.stream().filter(b -> b.aggregations() != null).map(b -> b.aggregations()).collect(Collectors.toList())
```

Comment:
Should we do this in all circumstances (i.e. in a `finally` block)?

(also it's ok to pass nulls to `Releasables#close`)

Reply:
I'd be fine with as much paranoia as you could manage. I'd probably move this to the top of the method to be honest.

I hadn't realized the aggregations were closeable - should we call `releaseAggs` anyway? That nulls the reference too I think.